staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go

```diff
@@ -0,0 +1,178 @@
+	VSphereInTreePluginName = "kubernetes.io/vsphere-volume"
+	// Name of parameter to supply SPBM Policy name for Volume Provisioning
+	paramStoragePolicyName = "storagepolicyname"
+	// Name of parameter to let controller know that it needs to return volumefilePath as Volume Attribute
+	// When CSI Driver Provision Volume using in-tree Storage Class and PVC
+	paramRequestVolumeFilePath   = "requestvolumefilepath"
+	paramVolumeFilePathAttribute = "volumefilepath"
+)
+
+var _ InTreePlugin = &vSphereCSITranslator{}
+
+// vSphereCSITranslator handles translation of PV spec from In-tree vSphere Volume to vSphere CSI
+type vSphereCSITranslator struct{}
+
+// NewvSphereCSITranslator returns a new instance of vSphereCSITranslator
+func NewvSphereCSITranslator() InTreePlugin {
+	return &vSphereCSITranslator{}
+}
+
+// TranslateInTreeStorageClassToCSI translates InTree vSphere storage class parameters to CSI storage class
+func (t *vSphereCSITranslator) TranslateInTreeStorageClassToCSI(sc *storage.StorageClass) (*storage.StorageClass, error) {
+	if sc == nil {
+		return nil, fmt.Errorf("sc is nil")
+	}
+	var params = map[string]string{}
+	// only "storagepolicyname" SC parameter in vsphere in-tree volume is
+	// supported by vSphere CSI Driver.
+	for k, v := range sc.Parameters {
+		paramName := strings.ToLower(k)
+		if paramName == paramStoragePolicyName {
+			params[paramStoragePolicyName] = v
+		} else if paramName == fsTypeKey {
+			// translate deprecated storage class parameter key fstype to csi.storage.k8s.io/fstype
+			params[csiFsTypeKey] = v
+		} else {
+			klog.V(2).Infof("StorageClass parameter [name:%q, value:%q] is not supported by the vSphere CSI driver. "+
+				"Translated Storage Class will not have this parameter.", paramName, v)
+		}
+	}
+	// sc.AllowedTopologies for Topology based volume provisioning will be supplied as it is.
+
+	// requestvolumefilepath param will help CSI Controller identify whether it needs to make a call to
+	// fetch vmdk path after Volume is created.
+	params[paramRequestVolumeFilePath] = "true"
+	sc.Parameters = params
+	return sc, nil
+}
+
+// TranslateInTreeInlineVolumeToCSI takes a Volume with VsphereVolume set from in-tree
+// and converts the VsphereVolume source to a CSIPersistentVolumeSource
+func (t *vSphereCSITranslator) TranslateInTreeInlineVolumeToCSI(volume *v1.Volume) (*v1.PersistentVolume, error) {
+	if volume == nil || volume.VsphereVolume == nil {
+		return nil, fmt.Errorf("volume is nil or VsphereVolume not defined on volume")
+	}
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			// Must be unique per disk as it is used as the unique part of the
+			// staging path
+			Name: fmt.Sprintf("%s-%s", VSphereDriverName, volume.VsphereVolume.VolumePath),
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				CSI: &v1.CSIPersistentVolumeSource{
+					Driver:           VSphereDriverName,
+					VolumeHandle:     volume.VsphereVolume.VolumePath,
```

Comment:
IIUC, the vSphere CSI driver will be modified to handle volume path for attach, detach, expand and delete, right?

Reply:
yes driver will be modified to handle volumepath for attach, detach and delete operation.